### PR TITLE
Porting all the TG atmos stuff (TESTMERGE READY)

### PR DESCRIPTION
**Pull request by Putnam3145**
## About The Pull Request

Ports https://github.com/tgstation/tgstation/pull/53013
Ports https://github.com/tgstation/tgstation/pull/50510
Ports https://github.com/tgstation/tgstation/pull/49821
Ports https://github.com/tgstation/tgstation/pull/43560
Ports https://github.com/tgstation/tgstation/pull/43362
Updates extools to have support for max temperatures on reactions
Updates extools to use a map instead of a vector, cause there's enough gases now that that's faster
Updates extools to check major gas, cause there's enough reactions/gases now that that's faster
Adds methane fires. The description on the can says it's flammable, now it is.
Makes nitryl and hyper-nob a good deal easier to make, by lowering their temperature requirements massively (though they still require large amounts of energy).

## Why It's Good For The Game

It's fast now, I'm using it.

## Changelog
:cl:
add: New Gas Reaction, Stimulum Energy Balls. Combine 100 degree Pluoxium, Stimulum and small amounts of Plasma and Nitryl to shoot off a deadly radioactive energy ball. The reaction will consume the plasma, pluoxium and stimulum and release small amounts of heat. The presence of water vapor will change the direction in which it shoots. See if you can figure out how to aim it.
add: Methane fires. Methane now reacts with oxygen and makes fire, generating CO2 and H2O.
tweak: Miasma sterilization now generates methane instead of oxygen.
add: N2O Decomposition reaction
tweak: N2O is no longer the perfect "safety gas". N2O Decomposes starting from 1400K, feeding more oxygen into the engine.
add: Added Freon
add: Added Hot Ice
add: Added interactions of freon with SM and reactions with other gases
add: Added H2 gas
add: Added H2 gas burn (reaction with o2, similar to tritium but without the radiations)
add: Added Electrolyzer (machine to make H2, requires power cell, once turned on and with h2o in the air will start converting it to h2 and o2 with the reaction 2H2O -> 2H2 + O2. Upgrading the parts will increase the amount of converted H2O and the efficency of the cell)
add: Added H2 interaction with SM (similar to trit, less radiation emission, same heat production and a bit of heat protection)
add: H2 reaction with N2 to make ammonia crystals (reaction in open air to make the crystals, temperatures between 0 °C and - 120° C, more N2 than H2)
add: H2 reaction with bz as catalyst at high heat and pressure (over 1e6 for both) to make metal hydrogen
add: metallic hydrogen golem (one of endish atmosian goal)
tweak: changed fusion to use h2 instead of co2, everything else is still the same
tweak: H2O vapor now slip under 30 °C so that plasma fires are not hellish
imageadd: added electrolyzer sprites (reuse of some unused ones)
imageadd: added ammonia crystals sprites
remove: fusion no longer generate stim and nitryl
tweak: Hyper-nob and nitryl both require energy to form instead of temperature.
balance: Hyper-nob and nitryl are easier to make due to this change.
add: hydrogen + rads = trit
add: Healium : made from Bz + Freon + cold, it knocks people down but have healing properties
add: Proto Nitrate : made from pluox and H2 + heat, it has different reaction with a bunch of other gases
add: Zauker: made from hypernoblium + stimulum around 0°C, it has heavy damage if breathed but is decomposed rapidly by N2
add: Halon : made from bz + trit + cold, it reacts with O2 at high temperature removing it and cooling the air, helping putting out fires
add: Hexane : made from bz + H2 + cold, it reacts with plasma and n2o and removes them from the air mix, filtering them.
tweak: stimulum no longer require plasma
add: canister sprites, gas sprites
add: crystal from gases formation and reaction
add: new metallic hydrogen recipes
add: elder atmosian statue
add: supermatter interactions
add : pluoxium can be also made without using rads (not removing the rad production one) (this also produces a minimal amount of h2)
add: zauker bounty
add: all 5 gases are up to selling
/:cl:

### CODE REVIEW

File: code/modules/surgery/organs/lungs.dm [hunk 1]
```diff
@@ -399,13 +403,67 @@
...
 			var/existing = H.reagents.get_reagent_amount(/datum/reagent/stimulum)
 			H.reagents.add_reagent(/datum/reagent/stimulum, max(0, 5 - existing))
 		breath.adjust_moles(/datum/gas/stimulum, -gas_breathed)
 
+	// Hyper-Nob
+		gas_breathed = breath.get_moles(/datum/gas/hypernoblium)
+		if (gas_breathed > gas_stimulation_min)
+			var/existing = H.reagents.get_reagent_amount(/datum/reagent/hypernoblium)
+			H.reagents.add_reagent(/datum/reagent/hypernoblium,max(0, 1 - existing))
+		breath.breath.adjust_moles(/datum/gas/hypernoblium,-gas_breathed)
```
Comment (Putnam3145) [inline code/modules/surgery/organs/lungs.dm]: wip

File: code/game/objects/items/grenades/atmos_grenades.dm [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+			continue
+		var/distance_from_center = max(get_dist(turf_loc, loc), 1)
+		var/turf/open/floor_loc = turf_loc
+		if(floor_loc.air.return_temperature() > 260 && floor_loc.air.return_temperature() < 370)
+			floor_loc.atmos_spawn_air("n2=[(gas_amount - 150) / distance_from_center];TEMP=273")
+		if(floor_loc.air.return_temperature() > 370)
+			floor_loc.atmos_spawn_air("n2=[gas_amount / distance_from_center];TEMP=30")
+			floor_loc.MakeSlippery(TURF_WET_PERMAFROST, (5 / distance_from_center) MINUTES)
+		if(floor_loc.air.get_moles(/datum/gas/plasma))
+			floor_loc.air.get_moles(/datum/gas/plasma) -= floor_loc.air.get_moles(/datum/gas/plasma) * 0.5 / distance_from_center
```
Comment (Putnam3145) [inline code/game/objects/items/grenades/atmos_grenades.dm]: wip

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 3]
```diff
@@ -1589,6 +1589,94 @@
...
+	name = "Healium"
+	description = "A Powerful sleeping agent with healing properties"
+	reagent_state = GAS
+	metabolization_rate = REAGENTS_METABOLISM * 0.5
+	color = "90560B"
+	taste_description = "rubber"
+
+/datum/reagent/healium/on_mob_metabolize(mob/living/L)
+	. = ..()
+	ADD_TRAIT(L, TRAIT_KNOCKEDOUT, type)
```
Comment (Putnam3145) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm]: wip

**Comment by Tupinambis** [comment]
Not sure I love Healium as a concept in general. It feels like a chem med doctors and antags can easily use to knock people out and get away with it since unlike other knockout gases, people aren't likely to question it.

Zauker just seems excessively potent, but I am not sure how easy/hard it is to make. Otherwise I welcome new gases.

**Comment by Putnam3145** [comment]
Zauker's requires hyper-nob and stimulum to make, both of which are pretty ridiculous.

Healium requires BZ and other weird stuff. I don't see doctors getting much of it.

**Comment by shellspeed1** [comment]
I'm going to be honest, I really am not a fan of Zauker, at least in regards to it exploding when mixed with proto nitrate. I can already see atmos mains blowing up entire stations by flooding the gas.

**Comment by Putnam3145** [comment]
> I can already see atmos mains blowing up entire stations by flooding the gas.

again, assuming they can make a ton of hyper-noblium and stimulum. unironically i'm probably the only person in the entire server who knows how to make stimulum properly.

**Comment by lightcat** [comment]
Gib now plz

**Comment by Putnam3145** [comment]
moving everything to rust sorry

**Comment by DeltaFire15** [comment]
Marking this as stale due to current nonmergable state aswell as not very recent commits.

**Comment by DeltaFire15** [comment]
Closing this PR due to staleness. Feel free to reopen once work on this PR is resumed.

**Pull request closed without merging**